<div style="text-align:left;font-size:2em"><span style="font-weight:bolder;font-size:1.25em">SP2273 | Learning Portfolio</span><br><br><span style="font-weight:bold;color:darkred">Loops (Good) Exercises</span></div>

# Exercise 1 (Make me an odd list)

In [30]:
%%timeit

happy=[]

for i in range(10):
    if i%2!=0:
        happy=happy+[i**2]
#print(happy)

2.31 μs ± 144 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


# Exercise 2 (Make me another odd list) ☻
List comprehension.

In [29]:
%%timeit
[number**2 for number in range(10) if number % 2 !=0]
#[output for i in range() if condition]

1.6 μs ± 115 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


# Exercise 3 (Time me!)
Cell magic command %%timeit to time the previous solutions. Which of the two is faster? **List comprehension**
\
**#1.** 2.31 μs ± 144 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)\
**#2.** 1.6 μs ± 115 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

# Exercise 4 (A problem of decay)

In [41]:
decayed = 0
undecayed = 100
counter=0

while undecayed>=decayed:
    decayed+=(0.05)*undecayed
    undecayed-=(0.05)*undecayed
    counter+=1

print(counter)

14


# Exercise 5 (Changes in CO2)

Identify those years that showed an increase of CO2 of 3 ppm or more compared to the previous year. Please print out these years along with the corresponding change in concentration.

In [123]:
co2_data = [
    (2000, 369.55), (2001, 371.14), (2002, 373.28), 
    (2003, 375.80), (2004, 377.52), (2005, 379.80), 
    (2006, 381.90), (2007, 383.79), (2008, 385.60), 
    (2009, 387.43), (2010, 389.90), (2011, 391.65), 
    (2012, 393.85), (2013, 396.52), (2014, 398.65),
    (2015, 400.83), (2016, 404.24), (2017, 406.55), 
    (2018, 408.52), (2019, 411.44), (2020, 414.24)
]

x=0
while x<len(co2_data)-1:
    if (co2_data[x][1]-co2_data[x+1][1])<=-3:
        print(f'{co2_data[x+1][0]},{abs(co2_data[x][1]-co2_data[x+1][1])}')
        x+=1
    else:
        x+=1

2016,3.410000000000025


### testing code

In [110]:
len(co2_data)

21

In [124]:
co2_data[15][1]-co2_data[15+1][1]

-3.410000000000025